### ISSUE TITLE
Improvment ideas

### ISSUE THREAD

**Issue by chrishoage** [chenxiaolong/ipmi-fan-control#10]
I've been using https://github.com/khorton/nas_fan_control for a while, and have always disliked that it was Perl. I had been thinking of writing my own in Rust to learn. Today I was poking around github and to my surprise found this, which is in the direction I wanted to go anyway. 

I have some ideas for future improvements and wanted to see what you think 

- Support for reading the HDD temperatures and averaging them as a sensor (either using smartctl or some other method of reading the data, along with a way to exclude drives)
- per zone interval. HDDs don't need to react as fast (say 10 seconds) as a CPU zone (say 1 second)
- curve support instead of steps. One thing I miss about a consumer motherboard is the ability to set a fan curve. Steps are pretty rigid, and a curve would allow for finer control and better ability to reduce fan noise. I imagine this would have the same data structure as the `steps` but instead of a upper and lower bound check the current temperature would control the duty cycle depending on where it fell between the curve steps

This project looks promising, once I find time to learn a little rust I may try to contribute my ideas if you want them in your project

**Comment by chenxiaolong** [comment]
Wow, glad you're interested in this project! I honestly didn't expect anyone else to even see the project. There can't be too many people who are using rackmount servers in a quiet environment and are also interested in Rust :)

> Support for reading the HDD temperatures and averaging them as a sensor (either using smartctl or some other method of reading the data, along with a way to exclude drives)

This is a good idea. I personally haven't done this because my SuperMicro 6028U-TR4T+'s two zones are for the two CPUs. There's no zone for just the disks. I ended up building [an external fan mount for the drives](https://user-images.githubusercontent.com/646253/68719038-fa31bf00-0578-11ea-853b-ef21d5b8f44f.jpg) that allows the internal fans to run much quieter without increasing the temperatures.

> per zone interval. HDDs don't need to react as fast (say 10 seconds) as a CPU zone (say 1 second)

This should be pretty easy to implement. I'd probably have the main loop to iterate at an interval that's the GCD of all the intervals specified in the config file. Each zone would just need to skip the necessary number of loop cycles.

> curve support instead of steps. One thing I miss about a consumer motherboard is the ability to set a fan curve. Steps are pretty rigid, and a curve would allow for finer control and better ability to reduce fan noise. I imagine this would have the same data structure as the steps but instead of a upper and lower bound check the current temperature would control the duty cycle depending on where it fell between the curve steps

ipmi-fan-control actually does this to some extent already. The steps define specific points on the "curve". The duty cycle for any temperature in between two steps is done via linear interpolation: https://github.com/chenxiaolong/ipmi-fan-control/blob/0d7a24fae5e3440f660c271cd5f8f6a301505306/src/main.rs#L182-L186 I'd be happy to implement different types of curves (besides linear) if it makes more sense though.

---

Another idea I'll probably work on in the near future is having ipmi-fan-control use libfreeipmi directly instead of communicating via `ipmitool shell`. My current code is [pretty ugly](https://github.com/chenxiaolong/ipmi-fan-control/blob/0d7a24fae5e3440f660c271cd5f8f6a301505306/src/ipmi.rs#L176-L220) and is dependent on ipmitool not changing how it prints out sensor readings.

**Comment by chrishoage** [comment]
> There can't be too many people who are using rackmount servers in a quiet environment and are also interested in Rust

There really can't! To be fair I haven't written any Rust yet, but have wanted to learn a strongly-typed language for a while now and Rust has intrigued me. This particular project was one I had picked to do myself, but I happened to find this first. 

If you didn't mind making this a linux only tool then you could already read CPU temperatures from `/sys/class/thermal/thermal_zone*/temp`  (at least on UEFI machines) seems to be pretty fast. Users would just need to define the thermal zone their CPU(s) are. Perhaps there could just be a config-level switch for faster linux performance or slower ipmi sensor readings. 

I'm running a X10SDV-6C+-TLN4F-O with a zone for CPU and a zone for HDD (I've only got two zones) 

When I was researching my project I didn't find a way to get HDD temps from anything other than S.M.A.R.T data from the drive. I couldn't quite figure out how `smartctl` queries data from the drives directly. I know `hddtemp` offers a server to poll data from, but it would be a bummer to introduce one more dependency on top of `ipmitool`. 

> The duty cycle for any temperature in between two steps is done via linear interpolation

That's great and basically what I was asking for anyway. That seems suitable for my needs. 

I might clone your project in the next couple months to see if I can hack in hdd temp monitoring from `hddtemp` to get started 

This looks great so far :slightly_smiling_face:

**Comment by chrishoage** [comment]
All these notes are mostly for myself. I've started researching this again, considering if I should fork this and learn rust or do something in python.

UEFI thermal zones seem like a good option for reading temperatures with no dependencies or shell calls. 

```
› cat /sys/class/thermal/thermal_zone1/type
x86_pkg_temp

chris at cortex in ~
› cat /sys/class/thermal/thermal_zone1/temp
40000

```

smartmontools v7 now supports returning json, which seems like the only option for retrieving hdd temp with out writing some sort of c binding. 

Config would need to accept a list of device paths, call smartctl and average the temperatures, optionally only considering the hottest `n` devices 

```
sudo smartctl  -l scttempsts -n standby  --json /dev/disk/by-id/ata-Crucial_CT525MX300SSD1_16431465A85A
{
  "json_format_version": [
    1,
    0
  ],
  "smartctl": {
    "version": [
      7,
      1
    ],
    "svn_revision": "5022",
    "platform_info": "x86_64-linux-5.4.0-54-generic",
    "build_info": "(local build)",
    "argv": [
      "smartctl",
      "-l",
      "scttempsts",
      "-n",
      "standby",
      "--json",
      "/dev/disk/by-id/ata-Crucial_CT525MX300SSD1_16431465A85A"
    ],
    "exit_status": 0
  },
  "device": {
    "name": "/dev/disk/by-id/ata-Crucial_CT525MX300SSD1_16431465A85A",
    "info_name": "/dev/disk/by-id/ata-Crucial_CT525MX300SSD1_16431465A85A [SAT]",
    "type": "sat",
    "protocol": "ATA"
  },
  "ata_sct_status": {
    "format_version": 3,
    "sct_version": 1,
    "device_state": {
      "value": 4,
      "string": "SMART Off-line Data Collection executing in background"
    },
    "temperature": {
      "current": 33,
      "power_cycle_min": 25,
      "power_cycle_max": 45,
      "lifetime_min": 18,
      "lifetime_max": 48,
      "under_limit_count": 0,
      "over_limit_count": 0
    }
  },
  "temperature": {
    "current": 33,
    "power_cycle_min": 25,
    "power_cycle_max": 45,
    "lifetime_min": 18,
    "lifetime_max": 48
  }
}
› sudo smartctl  -l scttempsts -n standby  --json /dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1X
{
  "json_format_version": [
    1,
    0
  ],
  "smartctl": {
    "version": [
      7,
      1
    ],
    "svn_revision": "5022",
    "platform_info": "x86_64-linux-5.4.0-54-generic",
    "build_info": "(local build)",
    "argv": [
      "smartctl",
      "-l",
      "scttempsts",
      "-n",
      "standby",
      "--json",
      "/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1X"
    ],
    "messages": [
      {
        "string": "Device is in STANDBY mode, exit(2)",
        "severity": "information"
      }
    ],
    "exit_status": 2
  },
  "device": {
    "name": "/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1X",
    "info_name": "/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1X [SAT]",
    "type": "sat",
    "protocol": "ATA"
  }
}

```

toml configuration could look like this 

```toml
[[zones]]
ipmi_zones = [0]
ipmi_fans=['FAN4']
interval=1
thermal_zones=[' /sys/class/thermal/thermal_zone1/temp']
steps = [
    { temp = 30, dcycle = 30 },
    { temp = 50, dcycle = 75 },
    { temp = 70, dcycle = 100 },
]

[[zones]]
ipmi_zones=[1]
ipmi_fans=['FAN1', 'FAN2', 'FAN3']
interval=10
smart_devices=[
  '/dev/disk/by-id/ata-Crucial_CT525MX300SSD1_16431465A85A',
  '/dev/disk/by-id/ata-Crucial_CT525MX300SSD1_16431465A85B',
  '/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1X',
  '/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1Y',
  '/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1Z',
]
npeek=4
steps = [
    { temp = 20, dcycle = 30 },
    { temp = 30, dcycle = 50 },
    { temp = 45, dcycle = 100 },
]

```

### New config: 

`ipmi_fans` is not 100% necessary but would be useful for logging in the zones, and offering sanity checks. (e.g. fan stalled, reset bmc and set to FULL). from `ipmitool sdr` 
`thermal_zones` is an array of fs path to a UEFI sys file that are averaged together and used for the temp for the steps
`smart_devices` is an array of devices to ask smartctl for temperatures about
`interval` is how frequently a zone should be considered for update 
`npeek` (just a random name, not tied to it) number of items in the array (either `thermal_zones` or `smart_devices`). This is useful so that a cooler drive than the rest doesn't bring the average down. 

### Removed config:

`sensors` stop relying on ipmitool to gather temperature data. 
`interval` each zone now declares an `interval`. Zone defaults to a safe default of 1s

This would be a quite a large change from what you currently have. If I were to do this in rust forking your project would you welcome the changes back upstream? 

If so, do you have implementation recommendations? I've (still) never touched rust so I'd be feeling my way though these changes. 

Thanks :slightly_smiling_face:

**Comment by chenxiaolong** [comment]
Thanks for the detailed writeup of your ideas!

> All these notes are mostly for myself. I've started researching this again, considering if I should fork this and learn rust or do something in python.

While I love Rust, I don't know if I'd recommend modifying this project as a first foray into Rust. In particular, the IPMI code is kind of awful. It launches `ipmitool shell` and parses the human-readable output to avoid unnecessary shell calls and IPMI reconnections. I feel like python with the python-ipmi module might be a better option.

> UEFI thermal zones seem like a good option for reading temperatures with no dependencies or shell calls.

I originally decided to only read the CPU temperature from IPMI so that ipmi-fan-control could be run from another host (or a VM on the same server). My own use case is for a server running Windows or some sort of appliance.

I'm not opposed to adding support for this as an option, but I'd prefer not to remove the ability to use IPMI sensors. (I realize reading HDD temperatures would necessitate running on the same host though.)

> smartmontools v7 now supports returning json, which seems like the only option for retrieving hdd temp with out writing some sort of c binding.

That is pretty awesome!

> toml configuration could look like this
> [removed]

I'd suggest maybe having a single eg. `sources` list where each item has the type. Maybe something like:

```toml
[[zones]]
ipmi_zones = [0]
interval = 1
sources = [
    { type = "file", path = "/sys/class/thermal/thermal_zone1/temp" },
    { type = "ipmi", sensor = "CPU2 Temp" },
]
steps = [
    { temp = 30, dcycle = 30 },
    { temp = 50, dcycle = 75 },
    { temp = 70, dcycle = 100 },
]

[[zones]]
ipmi_zones = [1]
interval = 10
sources = [
    { type = "smart", block_dev = "/dev/disk/by-id/ata-Crucial_CT525MX300SSD1_16431465A85A" },
    { type = "smart", block_dev = "/dev/disk/by-id/ata-Crucial_CT525MX300SSD1_16431465A85B" },
    { type = "smart", block_dev = "/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1X" },
    { type = "smart", block_dev = "/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1Y" },
    { type = "smart", block_dev = "/dev/disk/by-id/scsi-SATA_WDC_WD40EFRX-68W_WD-WCC4E4FKJH1Z" },
]
# Renamed from npeek (can't think of a good name either...)
sources_avg_top_n = 4
steps = [
    { temp = 20, dcycle = 30 },
    { temp = 30, dcycle = 50 },
    { temp = 45, dcycle = 100 },
]
```

> This would be a quite a large change from what you currently have. If I were to do this in rust forking your project would you welcome the changes back upstream?

Yes, I would! Actually, if you wouldn't mind, I'd be happy to implement some of the ideas you described here. I recently got a IPMI-enabled SuperMicro JBOD, so HDD temperature monitoring is now a top priority for me.

**Comment by chrishoage** [comment]
> While I love Rust, I don't know if I'd recommend modifying this project as a first foray into Rust. In particular, the IPMI code is kind of awful. It launches ipmitool shell and parses the human-readable output to avoid unnecessary shell calls and IPMI reconnections. I feel like python with the python-ipmi module might be a better option.

Well, if I'm being honest that was one of the reasons I considered forking this project, since it would offer a leg up with solving that particular problem 

>I originally decided to only read the CPU temperature from IPMI so that ipmi-fan-control could be run from another host (or a VM on the same server). My own use case is for a server running Windows or some sort of appliance.
>
> I'm not opposed to adding support for this as an option, but I'd prefer not to remove the ability to use IPMI sensors. (I realize reading HDD temperatures would necessitate running on the same host though.)

I did not consider this, and that makes total sense. The reason I was thinking about it was I'm not sure how much I trust the IPMI sensor compared to the package sensor. 

However, now that I have learned you scrape data though `ipmitool shell` it sounds more performant that I had originally thought. 

Perhaps this is something I could explore adding myself, and open a PR :slightly_smiling_face: 

> I'd suggest maybe having a single eg. sources list where each item has the type. Maybe something like:

I think that is a much better configuration option. It aligns the config between the zones allowing for the code to be made modular more easily.

> Renamed from npeek (can't think of a good name either...)
>sources_avg_top_n = 4

Names are hard :sweat_smile: 

This seems fine. It's not an entirely necessary configuration option, however I've found that with a large number of disks, particularly mixed 5400rpm and 7200rpm the temperature disparity between them can throw off the fan curve. 

For a minimum viable implementation of fan control for HDDs, we can actually just start with taking the "hottest" one and using that for the fan stepping temperature. You would sacrifice quieter fan speeds if only one disk is under load, however it would be a simpler first step. 

>Yes, I would! Actually, if you wouldn't mind, I'd be happy to implement some of the ideas you described here. I recently got a IPMI-enabled SuperMicro JBOD, so HDD temperature monitoring is now a top priority for me.

Of course I wouldn't mind at all. I am happy to help test. I've been eying this project for a year, but having a fan zone tied to HDD temperature was a requirement for me. 

I can explore opening a PR for some of the less useful options (reading temperature form a /sys file) especially if the bulk of the configuration is already changed to support the new format. 

Please let me know if there is any further information I can provide.

**Comment by chenxiaolong** [comment]
@chrishoage Could you give #15 a try? It adds support for the multiple sensor sources (IPMI, sysfs file, and smartctl). The temperature aggregation is still "max". I'll add support for avg in another PR once this one is merged.

**Comment by chrishoage** [comment]
That was incredibly fast! Thank you  for letting me know. I've left a review over in the PR. 

Everything seems to be working fine here!

**Comment by chenxiaolong** [comment]
Could you give #16 a try as well? It implements async loops for each zone, so slow source queries should no longer block other zones.

**Comment by chenxiaolong** [comment]
One final one: #17. This adds support for different source aggregation methods. I ended up going with a new `aggregation` option similar to how the sources are specified.

```toml
aggregation = { type = "maximum" }
aggregation = { type = "average" }
aggregation = { type = "average", top = 3 }
```

**Comment by chrishoage** [comment]
`aggregation` is much more clear, and leaves room for other aggregation types in the future. 

I uh... I think you can close this with that PR? This completely replaces the Perl IPMI fan control script I had modified, and there's no fiddly PID (a simple curve is so much easier to  understand). 

Is there anything I can do to contribute? Write unit test modules for you? Create a `Makefile` that builds and installs the binary, sample config, and systemd service? Figure out how to get auto-build-and-release on tags? 

I'd love to help out any way I can, you've just implemented all the features I had planned to try to do :sweat_smile:

**Comment by chenxiaolong** [comment]
Awesome! :D

I know you mentioned that you use Ubuntu on your server. Would you be interested in creating the necessary scripts for building a deb package? (for example, I have the scripts for creating a Fedora package in `dist/`).

I do really want to get the automatic builds and tagging working. I'm thinking of killing off the Azure Pipelines integration and switching to GitHub Actions similar to what I did for my other project: https://github.com/chenxiaolong/samfusdl/tree/master/.github/workflows

I'm thinking the workflow might be something like:

For the `master` branch and any PRs, just compile a debug executable and attach it as a GitHub Actions artifact.

For tags, build:
* Build `.src.rpm` for Fedora/CentOS
* Build whatever the Ubuntu packaging source format is
* Submit source package to Fedora Copr build service
* Submit source package to Ubuntu's Launchpad PPA build service
* Push PKGBUILD to AUR (maybe?)

**Comment by chrishoage** [comment]
>  Would you be interested in creating the necessary scripts for building a deb package? (for example, I have the scripts for creating a Fedora package in dist/).

Yeah, I've never made debpkg before but I'd love to try to figure it out. 

I'll also write up a PKGBUILD for Arch, should be pretty easy for this one. 

> I do really want to get the automatic builds and tagging working. I'm thinking of killing off the Azure Pipelines integration and switching to GitHub Actions similar to what I did for my other project

As far as the GitHub Actions, if you haven't been able to get around to porting over the workflow from your other project I'll take a look at that as well. 

A lot of this stuff I won't be able to get to until I have some time off for the holidays, however I'll be looking at the debpkg in the immediate future. 

Closing as complete. Thank you!
